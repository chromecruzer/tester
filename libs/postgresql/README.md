# @trac-postgresql

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test @trac-postgresql` to execute the unit tests via [Jest](https://jestjs.io).
