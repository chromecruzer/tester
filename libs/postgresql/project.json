{
  "name": "postgresql",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/postgresql/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/postgresql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/postgresql"],
      "options": {
        "jestConfig": "libs/postgresql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
