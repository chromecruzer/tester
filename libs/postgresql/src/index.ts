export * from './postgresql';
import PostgresConfigFactory, {PostgresConfig} from "./PostgresConfigFactory";
import BulkSqlWithCreate from "./BulkSqlWithCreate";
export *  from './ErrorHandler';
import ExtractIolData from "./ExtractIolData";
import PostgresClient from "./PostgresClient";
import LoadXlsxData from './LoadXlsxData';
import NotificationsServer from "./NotificationsServer";
import SearchTable from "./SearchTable";
import ImportIntoConsignments from "./ImportIntoConsignments";
import ImportIntoJumpTable from "./ImportIntoJumpTable";
import AuditStore from "./AuditStore";
import UserStore from './UserStore';

export {
  AuditStore, PostgresConfig, PostgresConfigFactory, BulkSqlWithCreate, ExtractIolData,
  PostgresClient, LoadXlsxData, NotificationsServer, SearchTable, ImportIntoConsignments, ImportIntoJumpTable,
  UserStore
};
