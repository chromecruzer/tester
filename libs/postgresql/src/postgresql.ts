import * as _ from "lodash";
import {DateTime} from "luxon";
import {PoolClient} from "pg";
import util from "util";
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import {
  customerContactDataMapping,
  customerDataMapping,
  dataTableNames, InputDataType,
  InputDataTypeLabel,
  InputToSqlMapping,
  iolDataMapping,
  matchCustomerContactFields,
  matchCustomerFields,
  matchIolFields,
  matchProductFields,
  matchSalesMappingFields,
  matchSalesRepFields,
  NullableString,
  productDataMapping,
  employeeRecordDm,
  salesMappingDataMapping,
  salesRepDataMapping,
  tracDateFormat, 
  UploadRecord,
  matchEmployeeFields,
  userRecordDm,
  matchUser1Fields
} from "../../datatypes/src";
import path from "path";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export const auditTableNames = {
  notes: 'auditnotes',
  locations: 'auditlocation',
  items: 'audititem',
}



export const searchTableNamesFn = () => [
  dataTableNames.consignments,
  dataTableNames.customers,
  dataTableNames.customerContacts,
  dataTableNames.products,
  dataTableNames.salesReps];

export const sqlDate = (datum: (Date | string)) => {
  // console.log(`datum ${dump(datum)}
  // from JSDate ${dump(DateTime.fromJSDate(datum))}
  // to format ${dump(DateTime.fromJSDate(datum).toFormat(tracDateFormat))}`)
  const date = _.isDate(datum) ? datum : new Date(datum);
  return DateTime.fromJSDate(date).toFormat(tracDateFormat)
}

export function datumToSql(datum: (string | number | Date | boolean), datatype: InputDataTypeLabel) {
  if (_.isNull(datum) || _.isUndefined(datum) || datum === '') {
    return 'NULL';
  }
  let str;
  switch (datatype) {
    case 'string':
      str = String(datum);
      if (str.length === 0) {
        return null;
      }
      str = str.replace(/'/g, "''");
      str = str.replace(/\n/g, "\\n");
      return `'${str}'`;
    case 'date':
      // console.log(`date ${datum} converted ${sqlDate(datum as unknown as Date)}`)
      return `'${sqlDate(datum as Date)}'`
    case 'boolean':
      // console.log(`boolean ${datum} as ${Boolean(datum)}`)
      return Boolean(datum);
    case 'number':
      return datum;
  }
}

export function appendUuids(sql: string, uuids: string[]) {
  let skipped1st = false;
  let result = sql;
  uuids.forEach(id => {
    result += `${skipped1st ? ',' : ''}    \n'${id}'`;
    skipped1st = true;
  });
  result += ');';
  return result;
}

export const checkModified = modified => {
  // console.log(`Checking ${dump(modified)}`);
  if (modified.originalUuid !== modified.new.uuid) {
    console.error(`UUID mismatch found ${dump(modified)}`);
  }
}

export function appendBulkData(sql, bulkData: UploadRecord[],
                               columnMapping: InputToSqlMapping[],
                               createUuid = true,
                               hasUuid = false) {
  let skipped1stRow = false;
  let result = `${sql}\n`;
  bulkData.forEach((record) => {
    // console.log(`append bulk data ${dump(record)}`)
    const dataRow = [];
    if (createUuid) {
      dataRow.push('gen_random_uuid()')
    }
    if(hasUuid) {
      dataRow.push(datumToSql(record['uuid'], 'string'));
    }
    columnMapping.forEach((colMap) => {
      let datum = datumToSql(record[colMap.sqlLabel], colMap.dataType);
      if(!_.isUndefined(colMap.default) && datum === null) {
        datum = colMap.default;
      }
      dataRow.push(datum)
    });
    result += `${skipped1stRow ? ',': ''}\n(${_.join(dataRow, ', ')})`;
    skipped1stRow = true;
  });
  return result;
}

export const queryRecords = async (client: PoolClient, recordSql: string, idField: string,
                                   {
                                     uuid = null as NullableString,
                                     id = null as NullableString,
                                     ids = [] as string[],
                                   }) => {
  let sql = recordSql;
  const check = (_.isString(uuid) ? 1 : 0) + (_.isString(id) ? 1 : 0) + (_.isEmpty(ids) ? 0 : 1);
  if (check > 1) {
    throw new Error(`cannot set multiple parameters ${uuid}, ${id}, ${ids.length}`);
  }
  if (uuid) {
    sql += ` WHERE uuid='${uuid}'`;
  }
  if (id) {
    sql += ` WHERE ${idField}='${id}'`;
  }
  if (!_.isEmpty(ids)) {
    sql += ` WHERE ${idField} IN (${_.join(_.map(ids, i => `'${i}'`), ', ')})`;
  }
  sql += ';';
  // console.log(`query records sql '${sql}'`);
  return client.query(sql)
    .then(data => data.rows)
    .catch(err => {
    throw new Error(`Mistake in '${sql}' caused ${err.message}`);
  });
}

interface TransformerKeys {
  matchFields: (string | string[]);
  matchFn: (a: any, b: any) => boolean,
  mappings: InputToSqlMapping[];
  compareFieldList?: string[],
}

export const recordFromArray = (row: InputDataType[], mapping: InputToSqlMapping[]) => {
  return _.reduce(mapping, (accum, m,i) => {
    const r = row[i] || null;
    let value = _.isFunction(m.nullify) ? m.nullify(r) : r;
    if(_.has(m, 'default')) { //need to us _.has here because the default may be a boolean false
      // console.log(`checking default for ${dump(m)} with value ${value}`)
      value = value || m.default;
    }
    accum[m.sqlLabel] = value;
    return accum;
  }, {} as UploadRecord);
}
export const transformHelpers: { [id: string]: TransformerKeys } = {
  iol: {
    matchFields: 'lot',
    matchFn: matchIolFields,
    mappings: iolDataMapping,
  },
  customers: {
    matchFields: 'customer_code',
    matchFn: matchCustomerFields,
    mappings: customerDataMapping,
  },
  customer_contacts: {
    matchFields: ['full_name', 'ship_to_id'],
    matchFn: matchCustomerContactFields,
    mappings: customerContactDataMapping,
  },
  products: {
    matchFields: 'product_id',
    matchFn: matchProductFields,
    mappings: productDataMapping,
  },
  employee: {
     matchFields: 'emp_id',
     matchFn: matchEmployeeFields,          // added by shyam for testing
     mappings: employeeRecordDm,
  },

  users: {
    matchFields: 'username',
    matchFn:  matchUser1Fields,              // added by shyam for testing
    mappings: userRecordDm
  },
  salesmappings: {
    matchFields: 'customer_ship_to',
    matchFn: matchSalesMappingFields,
    mappings: salesMappingDataMapping,
    compareFieldList: ['customer_ship_to', 'postal', 'zipcode', 'primary_territory', 'primary_name', 'cos_territory',
      'cos_name', 'kae_territory', 'kae_name'],
  },
  salesreps: {
    matchFields: 'salesrep_id',
    matchFn: matchSalesRepFields,
    mappings: salesRepDataMapping,
  },
}
export const matchFieldFn = (fields, record) => {
  return _.isArray(fields) ? _.join(_.map(fields, f => record[f]), '|') : record[fields]
}
export const deepEqualFn = (c, m, compareFieldList = null) => {
  if (_.isArray(compareFieldList)) {
    return _.reduce(compareFieldList,
      (boolAccum, cf) => {
        if (!boolAccum) {
          return boolAccum;
        }
        // console.log(`comparing isequal field ${cf}, ${candidate[cf]} to ${m[cf]}`);
        return _.isEqual(c[cf], m[cf]);
      }, true);
  }
  return _.isEqual(m, c);
}
export const docxTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.docx' : 'expiredTemplate.docx'))
export const htmlTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.html' : 'expiredTemplate.html'))

