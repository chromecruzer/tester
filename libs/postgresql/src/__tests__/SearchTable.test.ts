import {SearchTable} from "@trac/postgresql";

describe('SearchTable', () => {
  let searchTable, mockClient, mockSearchSignature;
  beforeEach(() => {
    mockClient = {
      query: jest.fn().mockReturnValue(Promise.resolve({rows: ['result1', 'result2']}))
    };
    searchTable = new SearchTable('test-schema-name');
    mockSearchSignature = {
      createSearchRecord: jest.fn().mockReturnValueOnce('searchField1')
        .mockReturnValueOnce('searchField2'),
      getSearchFields: jest.fn().mockReturnValue(['searchField1', 'searchField2']),
    };
    searchTable.searchSignatures['filled-table-schema-name.filled-table-name'] = mockSearchSignature;
  });
  describe('#createTable', () => {
    it('should create a search table correctly', () => {
      return searchTable.createTable(mockClient)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        })
    });
  });
  describe('fillTable', () => {
    it('should fill a table with search records correctly', () => {
      return searchTable.fillTable('filled-table-schema-name', 'filled-table-name', mockClient)
        .then(() => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
        })
    });

  });
  describe('search', () => {
    it('should return correctly formatted search results', () => {
      const expectedResults = [
        {
          schema: 'tracschema',
          tablename: 'customers',
          uuid: 'uuid1'
        },
        {
          schema: 'tracschema',
          tablename: 'customers',
          uuid: 'uuid2'
        },
        {
          schema: 'tracschema',
          tablename: 'customers',
          uuid: 'uuid3'
        },
        {
          schema: 'tracschema',
          tablename: 'products',
          uuid: 'uuid5'
        },
      ];
      const expectedCustomers = [
        {
          uuid: 'uuid1',
          customer_code: 'RES',
          name: 'efvwef1 - Pro',
        },
        {
          uuid: 'uuid2',
          customer_code: 'X RES',
          name: 'efvwef2 - Pro',
        },
        {
          uuid: 'uuid3',
          customer_code: 'RES1',
          name: 'efvwef3 - Pro',
        },
      ];
      const expectedProducts = [
        {
          uuid: 'uuid5',
          product_id: 'RES',
          description: 'PRO',
        },
      ];
      mockClient.query.mockReturnValueOnce(Promise.resolve({rows: expectedResults}))
        .mockReturnValueOnce(Promise.resolve({rows: expectedCustomers}))
        .mockReturnValueOnce(Promise.resolve({rows: expectedProducts}));
      return searchTable.search('pro', {consignments: true, products: false}, mockClient)
        .then(data => {
          expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
          expect(mockClient.query.mock.calls[1][0]).toMatchSnapshot();
          expect(mockClient.query.mock.calls[2][0]).toMatchSnapshot();
          expect(data).toMatchSnapshot();
        });
    });
  });
});
