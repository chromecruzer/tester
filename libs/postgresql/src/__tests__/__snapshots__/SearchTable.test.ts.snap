// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchTable #createTable should create a search table correctly 1`] = `
"CREATE TABLE test-schema-name.searchCache
(uuid UUID, schema varchar(20), tableName varchar(20), document tsvector);"
`;

exports[`SearchTable fillTable should fill a table with search records correctly 1`] = `
"WITH searchTable AS (SELECT uuid, to_tsvector(
 coalesce(searchField1, '') || ' ' ||  coalesce(searchField2, '')) AS doc from filled-table-schema-name.filled-table-name)
    INSERT INTO test-schema-name.searchCache
    SELECT uuid, 'filled-table-schema-name', 'filled-table-name', doc from searchTable;"
`;

exports[`SearchTable search should return correctly formatted search results 1`] = `
"SELECT uuid, schema, tablename FROM test-schema-name.searchCache
WHERE document @@ to_tsquery('pro:*')  AND tablename IN ( 'consignments')
ORDER BY ts_rank(document, to_tsquery('pro:*'))
DESC LIMIT 1200;"
`;

exports[`SearchTable search should return correctly formatted search results 2`] = `"select * from tracschema.customers where uuid in ( 'uuid1', 'uuid2', 'uuid3');"`;

exports[`SearchTable search should return correctly formatted search results 3`] = `"select * from tracschema.products where uuid in ( 'uuid5');"`;

exports[`SearchTable search should return correctly formatted search results 4`] = `
Array [
  Object {
    "consignment_id": "RES",
    "field": "efvwef1 - Pro",
    "signature": Object {
      "address": undefined,
      "city": undefined,
      "customer_code": "RES",
      "dea_number": undefined,
      "hin_number": undefined,
      "name": "efvwef1 - Pro",
      "phone": undefined,
      "postal": undefined,
      "state": undefined,
    },
    "type": "customers",
    "uuid": "uuid1",
  },
  Object {
    "consignment_id": "X RES",
    "field": "efvwef2 - Pro",
    "signature": Object {
      "address": undefined,
      "city": undefined,
      "customer_code": "X RES",
      "dea_number": undefined,
      "hin_number": undefined,
      "name": "efvwef2 - Pro",
      "phone": undefined,
      "postal": undefined,
      "state": undefined,
    },
    "type": "customers",
    "uuid": "uuid2",
  },
  Object {
    "consignment_id": "RES1",
    "field": "efvwef3 - Pro",
    "signature": Object {
      "address": undefined,
      "city": undefined,
      "customer_code": "RES1",
      "dea_number": undefined,
      "hin_number": undefined,
      "name": "efvwef3 - Pro",
      "phone": undefined,
      "postal": undefined,
      "state": undefined,
    },
    "type": "customers",
    "uuid": "uuid3",
  },
  Object {
    "consignment_id": "RES",
    "field": "PRO",
    "signature": Object {
      "description": "PRO",
      "description_level2": undefined,
      "description_level3": undefined,
      "description_level4": undefined,
      "description_level5": undefined,
      "product_id": "RES",
      "product_id_level2": undefined,
      "product_id_level3": undefined,
      "product_id_level4": undefined,
      "product_id_level5": undefined,
    },
    "type": "products",
    "uuid": "uuid5",
  },
]
`;
